### ISSUE TITLE
Discrete Continuous Elements

### ISSUE THREAD

**Issue by bruttenberg** [p2t2/figaro#400]
We should make discrete versions of continuous elements. This can be accomplished by making a Discrete trait that we can mix into continuous elements that overrides generateValue and discretizes the value to some range/scheme specified by the user (or whole numbers if none specified).

Factored algorithms can naturally use this interface by calling generateRandomness in a loop and collecting discrete values and their probabilities.

**Comment by mhoward2718** [comment]
Don't we have Abstraction for doing this?

**Comment by gtakata** [comment]
For a factored algorithm wouldn’t you just need the discretized version of the continuous distribution? True, some distributions might need to do this by sampling, but many can do it directly (sometimes using tables). The problem for a factor would be the size, ie how fine is the discretization?

*********************
Glenn Takata
Charles River Analytics Inc.
617.491.3474 x625
www.cra.com<http://www.cra.com>

**Comment by bruttenberg** [comment]
Yes mIchael, this is similar to abstraction but lifted up to the element level.

Glenn, as you say, there are different ways of thinking about this. One could describe a discretization scheme as whole numbers, so if you had a normal distribution, you could just sample the values and truncate to a whole number. For things like Uniform between 0 and 1, a table could solve the problem as well. But sampling is a very general solution, and we don't need special handling to build the factors.
